python/mage/constraint_programming/vrp_cp_solver.py

```diff
@@ -0,0 +1,356 @@
+                self.no_vehicles,
+            ),
+            NoBacktrackingConstraint(self._model, self._edge_chosen_vars),
+        ]
+
+        self._initialize()
+        self._add_constraints()
+        self._add_objective()
+        self._add_options()
+
+    def solve(self):
+        self._model.solve()
+
+    def get_result(self) -> VRPResult:
+        return VRPResult(
+            [
+                VRPPath(
+                    key[0] if key[0] >= 0 else self.depot_index,
+                    key[1] if key[1] >= 0 else self.depot_index,
+                )
+                for key, var in self._edge_chosen_vars.items()
+                if int(var.value[0]) == 1
+            ]
+        )
+
+    def get_distance(self, edge: Tuple[int, int]) -> float:
+        node_from, node_to = edge
+
+        if any(
+            node in [self.SOURCE_INDEX, self.SINK_INDEX]
+            for node in [node_from, node_to]
+        ):
+            return 0
+
+        return self.distance_matrix[node_from][node_to]
+
+    def _initialize(self):
+        for node_index in range(len(self.distance_matrix)):
+            if node_index in self._location_node_ids:
+                self._initialize_location_node(node_index)
+
+    def _initialize_location_node(self, node_index: int):
+        self._time_vars[node_index] = self._model.Var(value=0, lb=0, integer=False)
+
+        # Initialize starting point and sinking point for every vehicle
+        self._add_variable((self.SOURCE_INDEX, node_index))
+        self._add_variable((node_index, self.SINK_INDEX))
+
+        # For every node, draw lengths from and to it, with duration of edges
+        out_vars = self._add_adjacent_output_edge_variables(node_index)
+        in_vars = self._add_adjacent_input_edge_variables(node_index)
+
+        # Either it was a beginning node, or a vehicle has visited it in the drive.
+        if len(out_vars) > 0:
+            self._model.Equation(
+                self._edge_chosen_vars[(node_index, self.SINK_INDEX)] + sum(out_vars)
+                == 1
+            )
+
+        if len(in_vars) > 0:
+            self._model.Equation(
+                self._edge_chosen_vars[(self.SOURCE_INDEX, node_index)] + sum(in_vars)
+                == 1
+            )
+
+    def _add_adjacent_output_edge_variables(
+        self, node_index: int
+    ) -> List[Tuple[int, int]]:
+        edges_vars = []
+
+        for adjacent_node in range(len(self.distance_matrix)):
+            if adjacent_node == self.depot_index:
+                continue
+
+            edge = (node_index, adjacent_node)
+            var = self._add_variable(edge)
+            edges_vars.append(var)
+
+        return edges_vars
+
+    def _add_adjacent_input_edge_variables(
+        self, node_index: int
+    ) -> List[Tuple[int, int]]:
+        edges_vars = []
+
+        for adjacent_node in range(len(self.distance_matrix)):
+            if adjacent_node == self.depot_index:
+                continue
```

Comment:
There are two possible types of labels in this VRP scenario? Everything that is not considered a depot label is considered a location?